p2p/pubsub/topicRunner.go

```diff
@@ -0,0 +1,227 @@
+package pubsub
+
+import (
+	"context"
+	"sync"
+
+	"github.com/harmony-one/abool"
+	libp2p_pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+)
+
+// topicRunner runs the message handlers on a specific topic.
+// Currently, a topic runner is combined with multiple PubSubHandlers.
+// TODO: Redesign the topics and decouple the message usage according to Handler so that
+//       one topic has only one handler.
+type topicRunner struct {
+	topic       Topic
+	pubSub      rawPubSub
+	topicHandle topicHandle
+	options     []libp2p_pubsub.ValidatorOpt
+
+	// all active handlers in the topic; handlerLock protected
+	handlers    []Handler
+	handlerLock sync.RWMutex
+
+	ctxCancel func() // since upper function calls are single-threaded, there is no race here.
+
+	metric   *psMetric
+	running  abool.AtomicBool
+	closed   abool.AtomicBool
+	stopC    chan struct{}
+	stoppedC chan struct{}
+	log      zerolog.Logger
+}
+
+func newTopicRunner(host *pubSubHost, topic Topic, handlers []Handler, options []libp2p_pubsub.ValidatorOpt) (*topicRunner, error) {
+	tr := &topicRunner{
+		topic:    topic,
+		pubSub:   host.pubSub,
+		handlers: handlers,
+		options:  options,
+		stoppedC: make(chan struct{}),
+		log:      host.log.With().Str("pubSubTopic", string(topic)).Logger(),
+	}
+
+	tr.metric = newPsMetric(tr, defaultMetricInterval)
+
+	var err error
+	tr.topicHandle, err = tr.pubSub.Join(tr.topic)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot join topic [%v]", tr.topic)
+	}
+	if err := tr.pubSub.RegisterTopicValidator(tr.topic, tr.validateMsg, tr.options...); err != nil {
+		return nil, errors.Wrapf(err, "cannot register topic validator [%v]", tr.topic)
+	}
+
+	return tr, nil
+}
+
+func (tr *topicRunner) start() (err error) {
+	if changed := tr.running.SetToIf(false, true); !changed {
+		return errTopicAlreadyRunning
+	}
+	if tr.closed.IsSet() {
+		return errTopicClosed
+	}
+	defer func() {
+		if err != nil {
+			tr.running.SetTo(false)
+		}
+	}()
+
+	sub, err := tr.topicHandle.Subscribe()
+	if err != nil {
+		return errors.Wrapf(err, "cannot subscribe topic [%v]", tr.topic)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	tr.ctxCancel = cancel
+
+	go tr.metric.run()
+	go tr.run(ctx, sub)
+
+	return
+}
+
+func (tr *topicRunner) validateMsg(ctx context.Context, peer PeerID, raw *libp2p_pubsub.Message) libp2p_pubsub.ValidationResult {
+	m := newMessage(raw)
+	handlers := tr.getHandlers()
+	vResults := make([]ValidateResult, 0, len(handlers))
+
+	for _, handler := range handlers {
```

Comment:
are you expecting multiple validator for one message? This can be low efficient as validator has to run on every single libp2p messages.

Reply:
Yes, there will be node handler validating node message and consensus handler handling consensus message. So basically, messages are only validated by one handler - Another one is always return `MsgAccept`